File: itx-java/itx-impl/src/main/java/io/github/chehsunliu/itx/impl/queue/sqs/SqsMessageQueueFactory.java
package io.github.chehsunliu.itx.impl.queue.sqs;

import static io.github.chehsunliu.itx.impl.Env.envInt;
import static io.github.chehsunliu.itx.impl.Env.requireEnv;

import io.github.chehsunliu.itx.contract.queue.MessageQueue;
import io.github.chehsunliu.itx.contract.queue.MessageQueueFactory;
import java.net.URI;
import lombok.RequiredArgsConstructor;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.SqsClientBuilder;

@RequiredArgsConstructor
public final class SqsMessageQueueFactory implements MessageQueueFactory {
  private final SqsClient client;
  private final int maxConcurrency;
  private final String controlStandardQueueUrl;
  private final String controlPremiumQueueUrl;
  private final String computeStandardQueueUrl;
  private final String computePremiumQueueUrl;

  public static SqsMessageQueueFactory fromEnv() {
    String endpoint = System.getenv("ITX_SQS_LOCAL_ENDPOINT_URL");
    SqsClientBuilder builder = SqsClient.builder();
    if (endpoint != null && !endpoint.isBlank()) {
      builder.endpointOverride(URI.create(endpoint));
    }
    return new SqsMessageQueueFactory(
        builder.build(),
        envInt("ITX_SQS_MAX_CONCURRENCY", 100),
        requireEnv("ITX_SQS_CONTROL_STANDARD_QUEUE_URL"),
        requireEnv("ITX_SQS_CONTROL_PREMIUM_QUEUE_URL"),
        requireEnv("ITX_SQS_COMPUTE_STANDARD_QUEUE_URL"),
        requireEnv("ITX_SQS_COMPUTE_PREMIUM_QUEUE_URL"));
  }

  @Override
  public MessageQueue createControlStandardQueue() {
    return new SqsMessageQueue(client, controlStandardQueueUrl, maxConcurrency);
  }

  @Override
  public MessageQueue createControlPremiumQueue() {
    return new SqsMessageQueue(client, controlPremiumQueueUrl, maxConcurrency);
  }

  @Override
  public MessageQueue createComputeStandardQueue() {
    return new SqsMessageQueue(client, computeStandardQueueUrl, maxConcurrency);
  }

  @Override
  public MessageQueue createComputePremiumQueue() {
    return new SqsMessageQueue(client, computePremiumQueueUrl, maxConcurrency);
  }
}

File: itx-kt/itx-impl/src/main/kotlin/io/github/chehsunliu/itx/impl/queue/sqs/SqsMessageQueueFactory.kt
package io.github.chehsunliu.itx.impl.queue.sqs

import io.github.chehsunliu.itx.contract.queue.MessageQueue
import io.github.chehsunliu.itx.contract.queue.MessageQueueFactory
import io.github.chehsunliu.itx.impl.envInt
import io.github.chehsunliu.itx.impl.requireEnv
import software.amazon.awssdk.services.sqs.SqsClient
import java.net.URI

class SqsMessageQueueFactory(
    private val client: SqsClient,
    private val maxConcurrency: Int,
    private val controlStandardQueueUrl: String,
    private val controlPremiumQueueUrl: String,
    private val computeStandardQueueUrl: String,
    private val computePremiumQueueUrl: String,
) : MessageQueueFactory {
    companion object {
        fun fromEnv(): SqsMessageQueueFactory {
            val endpoint = System.getenv("ITX_SQS_LOCAL_ENDPOINT_URL")
            val builder = SqsClient.builder()
            if (!endpoint.isNullOrBlank()) builder.endpointOverride(URI.create(endpoint))
            return SqsMessageQueueFactory(
                client = builder.build(),
                maxConcurrency = envInt("ITX_SQS_MAX_CONCURRENCY", 100),
                controlStandardQueueUrl = requireEnv("ITX_SQS_CONTROL_STANDARD_QUEUE_URL"),
                controlPremiumQueueUrl = requireEnv("ITX_SQS_CONTROL_PREMIUM_QUEUE_URL"),
                computeStandardQueueUrl = requireEnv("ITX_SQS_COMPUTE_STANDARD_QUEUE_URL"),
                computePremiumQueueUrl = requireEnv("ITX_SQS_COMPUTE_PREMIUM_QUEUE_URL"),
            )
        }
    }

    override fun createControlStandardQueue(): MessageQueue = SqsMessageQueue(client, controlStandardQueueUrl, maxConcurrency)

    override fun createControlPremiumQueue(): MessageQueue = SqsMessageQueue(client, controlPremiumQueueUrl, maxConcurrency)

    override fun createComputeStandardQueue(): MessageQueue = SqsMessageQueue(client, computeStandardQueueUrl, maxConcurrency)

    override fun createComputePremiumQueue(): MessageQueue = SqsMessageQueue(client, computePremiumQueueUrl, maxConcurrency)
}
